File: binding/include/libqstatusbar.zig
const QtC = @import("qt6zig");
const qtc = @import("qt6c");%%_IMPORTLIBS_%%
%%_STRUCTDEFS_%%

/// https://doc.qt.io/qt-6/qstatusbar.html
pub const qstatusbar = struct {

    /// New constructs a new QStatusBar object.
    pub fn New(parent: ?*anyopaque) QtC.QStatusBar {
        return qtc.QStatusBar_new(@ptrCast(parent));
    }


    /// New2 constructs a new QStatusBar object.
    pub fn New2() QtC.QStatusBar {
        return qtc.QStatusBar_new2();
    }


    pub fn Tr(s: []const u8, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QStatusBar_Tr(@ptrCast(s));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qstatusbar.Tr: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn AddWidget(self: ?*anyopaque, widget: ?*anyopaque) void {
        qtc.QStatusBar_AddWidget(@ptrCast(self), @ptrCast(widget));
    }

    pub fn InsertWidget(self: ?*anyopaque, index: i32, widget: ?*anyopaque) i32 {
        return qtc.QStatusBar_InsertWidget(@ptrCast(self), index, @ptrCast(widget));
    }

    pub fn AddPermanentWidget(self: ?*anyopaque, widget: ?*anyopaque) void {
        qtc.QStatusBar_AddPermanentWidget(@ptrCast(self), @ptrCast(widget));
    }

    pub fn InsertPermanentWidget(self: ?*anyopaque, index: i32, widget: ?*anyopaque) i32 {
        return qtc.QStatusBar_InsertPermanentWidget(@ptrCast(self), index, @ptrCast(widget));
    }

    pub fn RemoveWidget(self: ?*anyopaque, widget: ?*anyopaque) void {
        qtc.QStatusBar_RemoveWidget(@ptrCast(self), @ptrCast(widget));
    }

    pub fn SetSizeGripEnabled(self: ?*anyopaque, sizeGripEnabled: bool) void {
        qtc.QStatusBar_SetSizeGripEnabled(@ptrCast(self), sizeGripEnabled);
    }

    pub fn IsSizeGripEnabled(self: ?*anyopaque, ) bool {
        return qtc.QStatusBar_IsSizeGripEnabled(@ptrCast(self));
    }

    pub fn CurrentMessage(self: ?*anyopaque, , allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QStatusBar_CurrentMessage(@ptrCast(self));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qstatusbar.CurrentMessage: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn ShowMessage(self: ?*anyopaque, text: []const u8) void {
        const text_str = qtc.libqt_string{
    .len = text.len,
    .data = text.ptr,
};
qtc.QStatusBar_ShowMessage(@ptrCast(self), text_str);
    }

    pub fn ClearMessage(self: ?*anyopaque, ) void {
        qtc.QStatusBar_ClearMessage(@ptrCast(self));
    }

    pub fn MessageChanged(self: ?*anyopaque, text: []const u8) void {
        const text_str = qtc.libqt_string{
    .len = text.len,
    .data = text.ptr,
};
qtc.QStatusBar_MessageChanged(@ptrCast(self), text_str);
    }

    pub fn OnMessageChanged(self: ?*anyopaque, callback: *const fn (?*anyopaque, [*:0]const u8) callconv(.c) void) void {
        qtc.QStatusBar_Connect_MessageChanged(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    pub fn ShowEvent(self: ?*anyopaque, param1: ?*anyopaque) void {
        qtc.QStatusBar_ShowEvent(@ptrCast(self), @ptrCast(param1));
    }

    /// Allows for overriding the related default method
    pub fn OnShowEvent(self: ?*anyopaque, callback: *const fn (?*anyopaque, ?*anyopaque) callconv(.c) void) void {
        qtc.QStatusBar_OnShowEvent(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    /// Base class method implementation
    pub fn QBaseShowEvent(self: ?*anyopaque, param1: ?*anyopaque) void {
        qtc.QStatusBar_QBaseShowEvent(@ptrCast(self), @ptrCast(param1));
    }

    pub fn PaintEvent(self: ?*anyopaque, param1: ?*anyopaque) void {
        qtc.QStatusBar_PaintEvent(@ptrCast(self), @ptrCast(param1));
    }

    /// Allows for overriding the related default method
    pub fn OnPaintEvent(self: ?*anyopaque, callback: *const fn (?*anyopaque, ?*anyopaque) callconv(.c) void) void {
        qtc.QStatusBar_OnPaintEvent(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    /// Base class method implementation
    pub fn QBasePaintEvent(self: ?*anyopaque, param1: ?*anyopaque) void {
        qtc.QStatusBar_QBasePaintEvent(@ptrCast(self), @ptrCast(param1));
    }

    pub fn ResizeEvent(self: ?*anyopaque, param1: ?*anyopaque) void {
        qtc.QStatusBar_ResizeEvent(@ptrCast(self), @ptrCast(param1));
    }

    /// Allows for overriding the related default method
    pub fn OnResizeEvent(self: ?*anyopaque, callback: *const fn (?*anyopaque, ?*anyopaque) callconv(.c) void) void {
        qtc.QStatusBar_OnResizeEvent(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    /// Base class method implementation
    pub fn QBaseResizeEvent(self: ?*anyopaque, param1: ?*anyopaque) void {
        qtc.QStatusBar_QBaseResizeEvent(@ptrCast(self), @ptrCast(param1));
    }

    pub fn Reformat(self: ?*anyopaque, ) void {
        qtc.QStatusBar_Reformat(@ptrCast(self));
    }

    /// Allows for overriding the related default method
    pub fn OnReformat(self: ?*anyopaque, callback: *const fn (?*anyopaque) callconv(.c) void) void {
        qtc.QStatusBar_OnReformat(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    /// Base class method implementation
    pub fn QBaseReformat(self: ?*anyopaque, ) void {
        qtc.QStatusBar_QBaseReformat(@ptrCast(self));
    }

    pub fn HideOrShow(self: ?*anyopaque, ) void {
        qtc.QStatusBar_HideOrShow(@ptrCast(self));
    }

    /// Allows for overriding the related default method
    pub fn OnHideOrShow(self: ?*anyopaque, callback: *const fn (?*anyopaque) callconv(.c) void) void {
        qtc.QStatusBar_OnHideOrShow(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    /// Base class method implementation
    pub fn QBaseHideOrShow(self: ?*anyopaque, ) void {
        qtc.QStatusBar_QBaseHideOrShow(@ptrCast(self));
    }

    pub fn Event(self: ?*anyopaque, param1: ?*anyopaque) bool {
        return qtc.QStatusBar_Event(@ptrCast(self), @ptrCast(param1));
    }

    /// Allows for overriding the related default method
    pub fn OnEvent(self: ?*anyopaque, callback: *const fn (?*anyopaque, ?*anyopaque) callconv(.c) bool) void {
        qtc.QStatusBar_OnEvent(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    /// Base class method implementation
    pub fn QBaseEvent(self: ?*anyopaque, param1: ?*anyopaque) bool {
        return qtc.QStatusBar_QBaseEvent(@ptrCast(self), @ptrCast(param1));
    }

    pub fn Tr2(s: []const u8, c: []const u8, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QStatusBar_Tr2(@ptrCast(s), @ptrCast(c));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qstatusbar.Tr2: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn Tr3(s: []const u8, c: []const u8, n: i32, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QStatusBar_Tr3(@ptrCast(s), @ptrCast(c), n);
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qstatusbar.Tr3: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn AddWidget2(self: ?*anyopaque, widget: ?*anyopaque, stretch: i32) void {
        qtc.QStatusBar_AddWidget2(@ptrCast(self), @ptrCast(widget), stretch);
    }

    pub fn InsertWidget3(self: ?*anyopaque, index: i32, widget: ?*anyopaque, stretch: i32) i32 {
        return qtc.QStatusBar_InsertWidget3(@ptrCast(self), index, @ptrCast(widget), stretch);
    }

    pub fn AddPermanentWidget2(self: ?*anyopaque, widget: ?*anyopaque, stretch: i32) void {
        qtc.QStatusBar_AddPermanentWidget2(@ptrCast(self), @ptrCast(widget), stretch);
    }

    pub fn InsertPermanentWidget3(self: ?*anyopaque, index: i32, widget: ?*anyopaque, stretch: i32) i32 {
        return qtc.QStatusBar_InsertPermanentWidget3(@ptrCast(self), index, @ptrCast(widget), stretch);
    }

    pub fn ShowMessage2(self: ?*anyopaque, text: []const u8, timeout: i32) void {
        const text_str = qtc.libqt_string{
    .len = text.len,
    .data = text.ptr,
};
qtc.QStatusBar_ShowMessage2(@ptrCast(self), text_str, timeout);
    }

    /// Delete this object from C++ memory.
    pub fn QDelete(self: ?*anyopaque) void {
        qtc.QStatusBar_Delete(@ptrCast(self));
    }
};

File: binding/src/libqstatusbar.cpp
#include <QEvent>
#include <QPaintEvent>
#include <QResizeEvent>
#include <QShowEvent>
#include <QStatusBar>
#include <QString>
#include <QByteArray>
#include <cstring>
#include <QWidget>
#include <qstatusbar.h>
#include "libqstatusbar.h"
#include "libqstatusbar.hxx"

QStatusBar* QStatusBar_new(QWidget* parent) {
	 return new VirtualQStatusBar(parent);
}

QStatusBar* QStatusBar_new2() {
	 return new VirtualQStatusBar();
}

libqt_string QStatusBar_Tr(const char* s) {
	QString _ret = QStatusBar::tr(s);
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

void QStatusBar_AddWidget(QStatusBar* self, QWidget* widget) {
	self->addWidget(widget);
}

int QStatusBar_InsertWidget(QStatusBar* self, int index, QWidget* widget) {
	return self->insertWidget(index, widget);
}

void QStatusBar_AddPermanentWidget(QStatusBar* self, QWidget* widget) {
	self->addPermanentWidget(widget);
}

int QStatusBar_InsertPermanentWidget(QStatusBar* self, int index, QWidget* widget) {
	return self->insertPermanentWidget(index, widget);
}

void QStatusBar_RemoveWidget(QStatusBar* self, QWidget* widget) {
	self->removeWidget(widget);
}

void QStatusBar_SetSizeGripEnabled(QStatusBar* self, bool sizeGripEnabled) {
	self->setSizeGripEnabled(sizeGripEnabled);
}

bool QStatusBar_IsSizeGripEnabled(const QStatusBar* self) {
	return self->isSizeGripEnabled();
}

libqt_string QStatusBar_CurrentMessage(const QStatusBar* self) {
	QString _ret = self->currentMessage();
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

void QStatusBar_ShowMessage(QStatusBar* self, const libqt_string text) {
	self->showMessage(QString::fromUtf8(text.data, text.len));
}

void QStatusBar_ClearMessage(QStatusBar* self) {
	self->clearMessage();
}

void QStatusBar_MessageChanged(QStatusBar* self, const libqt_string text) {
	self->messageChanged(QString::fromUtf8(text.data, text.len));
}

void QStatusBar_Connect_MessageChanged(QStatusBar* self, intptr_t slot) {
    void (*slotFunc)(QStatusBar*, const char*) = reinterpret_cast<void (*)(QStatusBar*, const char*)>(slot);
    QStatusBar::connect(self, &QStatusBar::messageChanged, [self, slotFunc](const QString& text) {
	slotFunc(self, text);
    });
}

libqt_string QStatusBar_Tr2(const char* s, const char* c) {
	QString _ret = QStatusBar::tr(s, c);
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

libqt_string QStatusBar_Tr3(const char* s, const char* c, int n) {
	QString _ret = QStatusBar::tr(s, c, n);
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

void QStatusBar_AddWidget2(QStatusBar* self, QWidget* widget, int stretch) {
	self->addWidget(widget, stretch);
}

int QStatusBar_InsertWidget3(QStatusBar* self, int index, QWidget* widget, int stretch) {
	return self->insertWidget(index, widget, stretch);
}

void QStatusBar_AddPermanentWidget2(QStatusBar* self, QWidget* widget, int stretch) {
	self->addPermanentWidget(widget, stretch);
}

int QStatusBar_InsertPermanentWidget3(QStatusBar* self, int index, QWidget* widget, int stretch) {
	return self->insertPermanentWidget(index, widget, stretch);
}

void QStatusBar_ShowMessage2(QStatusBar* self, const libqt_string text, int timeout) {
	self->showMessage(QString::fromUtf8(text.data, text.len), timeout);
}

// Derived class handler implementation
void QStatusBar_ShowEvent(QStatusBar* self, QShowEvent* param1) {
	auto* vqstatusbar = dynamic_cast<VirtualQStatusBar*>(self);
	if (vqstatusbar && vqstatusbar->isVirtualQStatusBar) {
	vqstatusbar->showEvent(param1);
	} else {
	self->QStatusBar::showEvent(param1);
}
}

// Base class handler implementation
void QStatusBar_QBaseShowEvent(QStatusBar* self, QShowEvent* param1) {
	auto* vqstatusbar = dynamic_cast<VirtualQStatusBar*>(self);
	if (vqstatusbar && vqstatusbar->isVirtualQStatusBar) {
vqstatusbar->setQStatusBar_ShowEvent_IsBase(true);
	vqstatusbar->showEvent(param1);
}
}

// Auxiliary method to allow providing re-implementation
void QStatusBar_OnShowEvent(QStatusBar* self, intptr_t slot) {
	auto* vqstatusbar = dynamic_cast<VirtualQStatusBar*>(self);
	if (vqstatusbar && vqstatusbar->isVirtualQStatusBar) {
vqstatusbar->setQStatusBar_ShowEvent_Callback(reinterpret_cast<VirtualQStatusBar::QStatusBar_ShowEvent_Callback>(slot));
}
}

// Derived class handler implementation
void QStatusBar_PaintEvent(QStatusBar* self, QPaintEvent* param1) {
	auto* vqstatusbar = dynamic_cast<VirtualQStatusBar*>(self);
	if (vqstatusbar && vqstatusbar->isVirtualQStatusBar) {
	vqstatusbar->paintEvent(param1);
	} else {
	self->QStatusBar::paintEvent(param1);
}
}

// Base class handler implementation
void QStatusBar_QBasePaintEvent(QStatusBar* self, QPaintEvent* param1) {
	auto* vqstatusbar = dynamic_cast<VirtualQStatusBar*>(self);
	if (vqstatusbar && vqstatusbar->isVirtualQStatusBar) {
vqstatusbar->setQStatusBar_PaintEvent_IsBase(true);
	vqstatusbar->paintEvent(param1);
}
}

// Auxiliary method to allow providing re-implementation
void QStatusBar_OnPaintEvent(QStatusBar* self, intptr_t slot) {
	auto* vqstatusbar = dynamic_cast<VirtualQStatusBar*>(self);
	if (vqstatusbar && vqstatusbar->isVirtualQStatusBar) {
vqstatusbar->setQStatusBar_PaintEvent_Callback(reinterpret_cast<VirtualQStatusBar::QStatusBar_PaintEvent_Callback>(slot));
}
}

// Derived class handler implementation
void QStatusBar_ResizeEvent(QStatusBar* self, QResizeEvent* param1) {
	auto* vqstatusbar = dynamic_cast<VirtualQStatusBar*>(self);
	if (vqstatusbar && vqstatusbar->isVirtualQStatusBar) {
	vqstatusbar->resizeEvent(param1);
	} else {
	self->QStatusBar::resizeEvent(param1);
}
}

// Base class handler implementation
void QStatusBar_QBaseResizeEvent(QStatusBar* self, QResizeEvent* param1) {
	auto* vqstatusbar = dynamic_cast<VirtualQStatusBar*>(self);
	if (vqstatusbar && vqstatusbar->isVirtualQStatusBar) {
vqstatusbar->setQStatusBar_ResizeEvent_IsBase(true);
	vqstatusbar->resizeEvent(param1);
}
}

// Auxiliary method to allow providing re-implementation
void QStatusBar_OnResizeEvent(QStatusBar* self, intptr_t slot) {
	auto* vqstatusbar = dynamic_cast<VirtualQStatusBar*>(self);
	if (vqstatusbar && vqstatusbar->isVirtualQStatusBar) {
vqstatusbar->setQStatusBar_ResizeEvent_Callback(reinterpret_cast<VirtualQStatusBar::QStatusBar_ResizeEvent_Callback>(slot));
}
}

// Derived class handler implementation
bool QStatusBar_Event(QStatusBar* self, QEvent* param1) {
	auto* vqstatusbar = dynamic_cast<VirtualQStatusBar*>(self);
	if (vqstatusbar && vqstatusbar->isVirtualQStatusBar) {
	return vqstatusbar->event(param1);
	} else {
	return self->QStatusBar::event(param1);
}
}

// Base class handler implementation
bool QStatusBar_QBaseEvent(QStatusBar* self, QEvent* param1) {
	auto* vqstatusbar = dynamic_cast<VirtualQStatusBar*>(self);
	if (vqstatusbar && vqstatusbar->isVirtualQStatusBar) {
vqstatusbar->setQStatusBar_Event_IsBase(true);
	return vqstatusbar->event(param1);
}
}

// Auxiliary method to allow providing re-implementation
void QStatusBar_OnEvent(QStatusBar* self, intptr_t slot) {
	auto* vqstatusbar = dynamic_cast<VirtualQStatusBar*>(self);
	if (vqstatusbar && vqstatusbar->isVirtualQStatusBar) {
vqstatusbar->setQStatusBar_Event_Callback(reinterpret_cast<VirtualQStatusBar::QStatusBar_Event_Callback>(slot));
}
}

// Derived class handler implementation
void QStatusBar_Reformat(QStatusBar* self) {
	auto* vqstatusbar = dynamic_cast<VirtualQStatusBar*>(self);
	if (vqstatusbar && vqstatusbar->isVirtualQStatusBar) {
	vqstatusbar->reformat();
	} else {
	self->QStatusBar::reformat();
}
}

// Base class handler implementation
void QStatusBar_QBaseReformat(QStatusBar* self) {
	auto* vqstatusbar = dynamic_cast<VirtualQStatusBar*>(self);
	if (vqstatusbar && vqstatusbar->isVirtualQStatusBar) {
vqstatusbar->setQStatusBar_Reformat_IsBase(true);
	vqstatusbar->reformat();
}
}

// Auxiliary method to allow providing re-implementation
void QStatusBar_OnReformat(QStatusBar* self, intptr_t slot) {
	auto* vqstatusbar = dynamic_cast<VirtualQStatusBar*>(self);
	if (vqstatusbar && vqstatusbar->isVirtualQStatusBar) {
vqstatusbar->setQStatusBar_Reformat_Callback(reinterpret_cast<VirtualQStatusBar::QStatusBar_Reformat_Callback>(slot));
}
}

// Derived class handler implementation
void QStatusBar_HideOrShow(QStatusBar* self) {
	auto* vqstatusbar = dynamic_cast<VirtualQStatusBar*>(self);
	if (vqstatusbar && vqstatusbar->isVirtualQStatusBar) {
	vqstatusbar->hideOrShow();
	} else {
	self->QStatusBar::hideOrShow();
}
}

// Base class handler implementation
void QStatusBar_QBaseHideOrShow(QStatusBar* self) {
	auto* vqstatusbar = dynamic_cast<VirtualQStatusBar*>(self);
	if (vqstatusbar && vqstatusbar->isVirtualQStatusBar) {
vqstatusbar->setQStatusBar_HideOrShow_IsBase(true);
	vqstatusbar->hideOrShow();
}
}

// Auxiliary method to allow providing re-implementation
void QStatusBar_OnHideOrShow(QStatusBar* self, intptr_t slot) {
	auto* vqstatusbar = dynamic_cast<VirtualQStatusBar*>(self);
	if (vqstatusbar && vqstatusbar->isVirtualQStatusBar) {
vqstatusbar->setQStatusBar_HideOrShow_Callback(reinterpret_cast<VirtualQStatusBar::QStatusBar_HideOrShow_Callback>(slot));
}
}

void QStatusBar_Delete(QStatusBar* self) {
    delete self;
}

